File: go/player.go
package main

import (
	"fmt"
	"slices"
	"strings"
)

// PlayerState represents the current state of a player in a round
type PlayerState int

const (
	Active PlayerState = iota
	Stayed
	Busted
)

// PlayerType represents whether a player is human or computer
type PlayerType int

const (
	Human PlayerType = iota
	Computer
)

type PlayerInterface interface {
	GetName() string
	HasSecondChance() bool
	GetTotalScore() int
	GetPlayerIcon() string
	AddCard(card *Card) error
	UseSecondChance() *Card
	Stay()
	Bust()
	CalculateRoundScore() int
	AddToTotalScore()
	ResetForNewRound() []*Card
	IsActive() bool
	HasCards() bool
	ShowHand()
	GetHand() []*Card
	GetHandSummary() string
	ChooseActionTarget(gameState *GameState, actionType ActionType) (PlayerInterface, error)
	ChoosePositiveActionTarget(gameState *GameState, actionType ActionType) (PlayerInterface, error)
	MakeHitStayDecision(gameState *GameState) (bool, error)
}

// Player represents a game player
type BasePlayer struct {
	Name          string
	TotalScore    int
	NumberCards   []*Card
	ModifierCards []*Card
	ActionCards   []*Card
	State         PlayerState
	SecondChance  bool
}

func (p *BasePlayer) Init(name string) {
	p.Name = name
	p.NumberCards = make([]*Card, 0)
	p.ModifierCards = make([]*Card, 0)
	p.ActionCards = make([]*Card, 0)
	p.State = Active
}

func (p *BasePlayer) GetName() string {
	return p.Name
}

func (p *BasePlayer) HasSecondChance() bool {
	return p.SecondChance
}

func (p *BasePlayer) GetTotalScore() int {
	return p.TotalScore
}

// AddCard adds a card to the player's hand
func (p *BasePlayer) AddCard(card *Card) error {
	switch card.Type {
	case NumberCard:
		// Check for duplicate number cards (busting)
		for _, existing := range p.NumberCards {
			if existing.Value == card.Value {
				// Player busts unless they have a second chance
				if p.HasSecondChance() {
					return fmt.Errorf("duplicate_with_second_chance:%d", card.Value)
				}
				p.Bust()
				return fmt.Errorf("bust:%d", card.Value)
			}
		}
		p.NumberCards = append(p.NumberCards, card)

		// Check for Flip 7
		if len(p.NumberCards) == 7 {
			p.Stay()
			return fmt.Errorf("flip7")
		}

	case ModifierCard:
		p.ModifierCards = append(p.ModifierCards, card)

	case ActionCard:
		if card.Action == SecondChance {
			if p.HasSecondChance() {
				return fmt.Errorf("second_chance_duplicate")
			}
			p.SecondChance = true
		}
		p.ActionCards = append(p.ActionCards, card)
	}

	return nil
}

func (p *BasePlayer) GetHand() []*Card {
	return slices.Concat(p.NumberCards, p.ModifierCards, p.ActionCards)
}

// UseSecondChance uses the second chance card to avoid busting
func (p *BasePlayer) UseSecondChance() *Card {
	if !p.HasSecondChance() {
		panic("no second change card to use")
	}

	p.SecondChance = false

	// Remove second chance card
	for i, card := range p.ActionCards {
		if card.Action == SecondChance {
			p.ActionCards = append(p.ActionCards[:i], p.ActionCards[i+1:]...)
			return card
		}
	}

	panic("no second chance card to use")
}

// Stay makes the player stay and bank their points
func (p *BasePlayer) Stay() {
	if p.State == Active {
		p.State = Stayed
	} else {
		panic("stay when player is not active")
	}
}

func (p *BasePlayer) Bust() {
	if p.State == Active {
		p.State = Busted
	} else {
		panic("bust when player is not active")
	}
}

// CalculateRoundScore calculates the player's score for the current round
func (p *BasePlayer) CalculateRoundScore() int {
	if p.State == Busted {
		return 0
	}

	// Calculate base score from number cards
	numberTotal := 0
	for _, card := range p.NumberCards {
		numberTotal += card.Value
	}

	// Apply multiplier if present
	for _, card := range p.ModifierCards {
		if card.Modifier == Multiply2 {
			numberTotal *= 2
			break
		}
	}

	// Add modifier points
	modifierTotal := 0
	for _, card := range p.ModifierCards {
		if card.Modifier != Multiply2 {
			modifierTotal += card.GetPoints()
		}
	}

	total := numberTotal + modifierTotal

	// Add Flip 7 bonus
	if len(p.NumberCards) == 7 {
		total += 15
	}

	return total
}

// AddToTotalScore adds the round score to the total score
func (p *BasePlayer) AddToTotalScore() {
	p.TotalScore += p.CalculateRoundScore()
}

// ResetForNewRound resets the player's state for a new round
func (p *BasePlayer) ResetForNewRound() []*Card {
	discardedCards := p.GetHand()
	p.NumberCards = make([]*Card, 0)
	p.ModifierCards = make([]*Card, 0)
	p.ActionCards = make([]*Card, 0)
	p.State = Active
	p.SecondChance = false
	return discardedCards
}

// IsActive returns true if the player is still active in the current round
func (p *BasePlayer) IsActive() bool {
	return p.State == Active
}

// HasCards returns true if the player has any number cards
func (p *BasePlayer) HasCards() bool {
	return len(p.NumberCards) > 0
}

// ShowHand displays the player's current hand
func (p *BasePlayer) ShowHand() {
	fmt.Printf("%s:\n", p.Name)

	if len(p.NumberCards) == 0 && len(p.ModifierCards) == 0 {
		fmt.Println("   No cards")
		return
	}

	// Show number cards
	if len(p.NumberCards) > 0 {
		fmt.Print("   Numbers: ")
		for i, card := range p.NumberCards {
			if i > 0 {
				fmt.Print(" ")
			}
			fmt.Print(card.String())
		}
		fmt.Println()
	}

	// Show modifier cards
	if len(p.ModifierCards) > 0 {
		fmt.Print("   Modifiers: ")
		for i, card := range p.ModifierCards {
			if i > 0 {
				fmt.Print(" ")
			}
			fmt.Print(card.String())
		}
		fmt.Println()
	}

	// Show special status
	if p.HasSecondChance() {
		fmt.Println("   🆘 Has Second Chance")
	}

	// Show state
	switch p.State {
	case Stayed:
		fmt.Printf("   ✅ STAYED - Round Score: %d\n", p.CalculateRoundScore())
	case Busted:
		fmt.Println("   💥 BUSTED")
	}

	fmt.Println()
}

// GetHandSummary returns a compact summary of the player's hand
func (p *BasePlayer) GetHandSummary() string {
	if p.State == Busted {
		return "💥 BUSTED"
	}

	if len(p.NumberCards) == 0 && len(p.ModifierCards) == 0 {
		return "No cards"
	}

	parts := make([]string, 0)

	if len(p.NumberCards) > 0 {
		numbers := make([]string, len(p.NumberCards))
		for i, card := range p.NumberCards {
			numbers[i] = fmt.Sprintf("%d", card.Value)
		}
		parts = append(parts, strings.Join(numbers, ","))
	}

	if len(p.ModifierCards) > 0 {
		mods := make([]string, len(p.ModifierCards))
		for i, card := range p.ModifierCards {
			mods[i] = card.String()
		}
		parts = append(parts, strings.Join(mods, ","))
	}

	result := strings.Join(parts, " | ")

	if p.State == Stayed {
		result += fmt.Sprintf(" (STAYED: %d pts)", p.CalculateRoundScore())
	}

	return result
}

File: player.py
"""Player state and hand management for Flip 7."""

from __future__ import annotations

from enum import IntEnum
from typing import List, Optional

from card import Card, CardType, ActionType, ModifierType


class PlayerState(IntEnum):
    ACTIVE = 0
    STAYED = 1
    BUSTED = 2


class PlayerType(IntEnum):
    HUMAN = 0
    COMPUTER = 1


class GameState:
    """Context for AI decision making."""

    def __init__(
        self,
        round_num: int,
        players: List[BasePlayer],
        active_players: List[BasePlayer],
        current_leader: Optional[BasePlayer],
        cards_in_deck: List[Card],
    ) -> None:
        self.round = round_num
        self.players = players
        self.active_players = active_players
        self.current_leader = current_leader
        self.cards_in_deck = cards_in_deck


class BasePlayer:
    """Base player state and hand logic."""

    def __init__(self, name: str = "") -> None:
        self.name = name
        self.total_score = 0
        self.number_cards: List[Card] = []
        self.modifier_cards: List[Card] = []
        self.action_cards: List[Card] = []
        self.state = PlayerState.ACTIVE
        self.second_chance = False

    def init(self, name: str) -> None:
        self.name = name
        self.number_cards = []
        self.modifier_cards = []
        self.action_cards = []
        self.state = PlayerState.ACTIVE

    def get_name(self) -> str:
        return self.name

    def has_second_chance(self) -> bool:
        return self.second_chance

    def get_total_score(self) -> int:
        return self.total_score

    def add_card(self, card: Card) -> Optional[str]:
        """Add a card to the player's hand. Returns error string or None."""
        if card.type == CardType.NUMBER:
            for existing in self.number_cards:
                if existing.value == card.value:
                    if self.has_second_chance():
                        return f"duplicate_with_second_chance:{card.value}"
                    self.bust()
                    return f"bust:{card.value}"
            self.number_cards.append(card)
            if len(self.number_cards) == 7:
                self.stay()
                return "flip7"
            return None

        if card.type == CardType.MODIFIER:
            self.modifier_cards.append(card)
            return None
        if card.type == CardType.ACTION:
            if card.action == ActionType.SECOND_CHANCE:
                if self.has_second_chance():
                    return "second_chance_duplicate"
                self.second_chance = True
            self.action_cards.append(card)
            return None
        return None

    def get_hand(self) -> List[Card]:
        return self.number_cards + self.modifier_cards + self.action_cards

    def use_second_chance(self) -> Card:
        if not self.has_second_chance():
            raise RuntimeError("no second chance card to use")
        self.second_chance = False
        for i, card in enumerate(self.action_cards):
            if card.action == ActionType.SECOND_CHANCE:
                self.action_cards.pop(i)
                return card
        raise RuntimeError("no second chance card to use")

    def stay(self) -> None:
        if self.state == PlayerState.ACTIVE:
            self.state = PlayerState.STAYED
        else:
            raise RuntimeError("stay when player is not active")

    def bust(self) -> None:
        if self.state == PlayerState.ACTIVE:
            self.state = PlayerState.BUSTED
        else:
            raise RuntimeError("bust when player is not active")

    def calculate_round_score(self) -> int:
        if self.state == PlayerState.BUSTED:
            return 0
        number_total = sum(c.value for c in self.number_cards)
        for c in self.modifier_cards:
            if c.modifier == ModifierType.MULTIPLY_2:
                number_total *= 2
                break
        modifier_total = sum(
            c.get_points() for c in self.modifier_cards
            if c.modifier != ModifierType.MULTIPLY_2
        )
        total = number_total + modifier_total
        if len(self.number_cards) == 7:
            total += 15
        return total

    def add_to_total_score(self) -> None:
        self.total_score += self.calculate_round_score()

    def reset_for_new_round(self) -> List[Card]:
        discarded = self.get_hand()
        self.number_cards = []
        self.modifier_cards = []
        self.action_cards = []
        self.state = PlayerState.ACTIVE
        self.second_chance = False
        return discarded

    def is_active(self) -> bool:
        return self.state == PlayerState.ACTIVE

    def has_cards(self) -> bool:
        return len(self.number_cards) > 0

    def show_hand(self) -> None:
        print(f"{self.name}:")
        if not self.number_cards and not self.modifier_cards:
            print("   No cards")
            return
        if self.number_cards:
            print("   Numbers: ", " ".join(str(c) for c in self.number_cards))
        if self.modifier_cards:
            print("   Modifiers: ", " ".join(str(c) for c in self.modifier_cards))
        if self.has_second_chance():
            print("   🆘 Has Second Chance")
        if self.state == PlayerState.STAYED:
            print(f"   ✅ STAYED - Round Score: {self.calculate_round_score()}")
        if self.state == PlayerState.BUSTED:
            print("   💥 BUSTED")
        print()

    def get_hand_summary(self) -> str:
        if self.state == PlayerState.BUSTED:
            return "💥 BUSTED"
        if not self.number_cards and not self.modifier_cards:
            return "No cards"
        parts = []
        if self.number_cards:
            parts.append(",".join(str(c.value) for c in self.number_cards))
        if self.modifier_cards:
            parts.append(",".join(str(c) for c in self.modifier_cards))
        result = " | ".join(parts)
        if self.state == PlayerState.STAYED:
            result += f" (STAYED: {self.calculate_round_score()} pts)"
        return result
